syne_tune/optimizer/schedulers/searchers/bayesopt/contributed/predictor.py

```diff
@@ -0,0 +1,121 @@
+# Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License").
+# You may not use this file except in compliance with the License.
+# A copy of the License is located at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+from typing import Tuple, List, Dict, Optional
+
+import numpy as np
+
+from syne_tune.optimizer.schedulers.searchers.bayesopt.datatypes.tuning_job_state import (
+    TuningJobState,
+)
+from syne_tune.optimizer.schedulers.searchers.bayesopt.models.model_base import (
+    BasePredictor,
+)
+
+
+class ContributedPredictor:
+    """
+    Base class for the contributed predictors
+    """
+
+    def predict(
+        self, X: np.ndarray, return_std: bool = True
+    ) -> Tuple[np.ndarray, np.ndarray]:
+        """
+        Returns signals which are statistics of the predictive distribution at
+        input points ``inputs``.
+
+
+        :param inputs: Input points, shape ``(n, d)``
+        :return: Tuple with the following entries:
+            * "mean": Predictive means in shape of ``(n,)``
+            * "std": Predictive stddevs, shape ``(n,)``
+        """
+
+        raise NotImplementedError()
+
+    def backward_gradient(
+        self, input: np.ndarray, head_gradients: List[Dict[str, np.ndarray]]
+    ) -> np.ndarray:
+        """
+        Computes the gradient :math:`\nabla f(x)` for an acquisition
+        function :math:`f(x)`, where :math:`x` is a single input point. This
+        is using reverse mode differentiation, the head gradients are passed
+        by the acquisition function. The head gradients are
+        :math:`\partial_k f`, where :math:`k` runs over the statistics
+        returned by :meth:`predict` for the single input point :math:`x`.
+        The shape of head gradients is the same as the shape of the
+        statistics.
+
+        :param input: Single input point :math:`x`, shape ``(d,)``
+        :param head_gradients: See above
+        :return: Gradient :math:`\nabla f(x)`
+        """
+        raise NotImplementedError()
+
+
+class ContributedPredictorWrapper(BasePredictor):
```

Comment:
I think this one should move to `models/contributed_estimator.py` (except that "contributed" should become something else).

This is like for GP: The wrappers are in `bayesopt/models` (both estimator and predictor in the same wrapper code file), the code is in `bayesopt/gpautograd`

Reply:
Ok makes sense, will move.

Reply:
And note that your code in `bayesopt/contributed` need not adhere to the estimator / predictor split, but you can keep everything in a single class. The split into two is only required one level up, in the wrappers.

That's what I did for the GP code, and also what we discussed.

Reply:
Thats a good point but I wanted to keep them separated to make it more streamlined. This way we rely on the decoupled view through all the levels and I think it will be easier to follow for newer contributors. I know that I was confused by couplig/decoupling in GP code the first time I saw it.